controllers/cloudstackmachine_controller.go

```diff
@@ -223,11 +223,11 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateVMInstance() (retRes
 		r.Recorder.Eventf(r.ReconciliationSubject, "Warning", "Creating", CSMachineCreationFailed, err.Error())
 	}
 	if err == nil && !controllerutil.ContainsFinalizer(r.ReconciliationSubject, infrav1.MachineFinalizer) { // Fetched or Created?
+		controllerutil.AddFinalizer(r.ReconciliationSubject, infrav1.MachineFinalizer)
 		r.Recorder.Eventf(r.ReconciliationSubject, "Normal", "Created", CSMachineCreationSuccess)
 		r.Log.Info(CSMachineCreationSuccess, "instanceStatus", r.ReconciliationSubject.Status)
 	}
-	// Always add the finalizer regardless. It can't be added twice anyway.
-	controllerutil.AddFinalizer(r.ReconciliationSubject, infrav1.MachineFinalizer)
```

Comment:
would there ever be any resources under the cloudstackmachine which would need to be cleaned up? Another layer of abstraction? What if we start managing volumes independently of machines someday? Wouldn't a lack of a finalizer result in orphaned sub-resources? It's not a blocker, but I want to be sure we're confident in this approach

Reply:
I suspect it's impossible to predict exactly how a future developer will modify things.  If code to create the proposed sub-resources runs after VM creation, then I believe these will never be created, so all is well.  If such code is added before VM creation, then yes, these would be created, and could be orphaned at delete time.  Since this is non-obvious behavior, perhaps best we can do is add a highly visible comment (which should also mention the problem that occurs if the finalizer is created despite VM creation failure).

Reply:
If some other resource owned (created ) by cloudstackmachine, that owned resource will have an ownerRef to cloudstackmachine. Once cloudstackMachine is deleted, owned resource will be deleted too by kube -- there are several variants of this depends on deletion cascade policy.

Reply:
Fair enough.  I'd still like to see a comment to inform future devs about the reason for and implications of the lifecycle management technique coded.